src/App.tsx

```diff
@@ -1,24 +1,345 @@
+      filteredTodos = sourceTodos.filter(todo => todo.completed);
+      break;
+
+    default:
+      filteredTodos = [...sourceTodos];
   }
 
+  return filteredTodos;
+};
+
+export const App: React.FC = () => {
+  const [
+    selectedFilter,
+    setSelectedFilter,
+  ] = useState<FilterType>(FilterType.All);
+
+  const [todosFromServer, setTodosFromServer] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [selectedTodoIds, setSelectedTodoIds] = useState<number[]>([]);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const [
+    currentError,
+    setCurrentError,
+  ] = useState<ErrorType>(ErrorType.NoError);
+
+  const timerId = useRef<NodeJS.Timeout>();
+
+  const onError = (error: ErrorType) => {
+    if (timerId.current) {
+      clearTimeout(timerId.current);
+    }
+
+    setCurrentError(error);
+    timerId.current = setTimeout(
+      () => setCurrentError(ErrorType.NoError),
+      3000,
+    );
+  };
+
+  const handleSubmit = async (
+    event: FormEvent<HTMLFormElement>,
+    newTodoText: string,
+    setNewTodoText: Dispatch<SetStateAction<string>>,
+  ) => {
+    event.preventDefault();
+
+    if (!newTodoText.trim()) {
+      onError(ErrorType.EmptyTitle);
+
+      return;
+    }
+
+    const newTodo = {
+      id: 0,
+      userId: USER_ID,
+      title: newTodoText.trim(),
+      completed: false,
+    };
+
+    try {
+      setTempTodo(() => newTodo);
+      await addTodo(newTodo)
+        .then((todo) => setTodosFromServer(
+          prevTodos => ([...prevTodos, todo]),
+        ))
+        .then(() => {
+          setNewTodoText('');
+        });
+    } catch {
+      onError(ErrorType.UnableToAdd);
+    } finally {
+      setTempTodo(() => null);
+
+      setTimeout(() => {
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      }, 100);
+    }
+  };
+
+  const handleSingleTodoDelition = async (todoId: number) => {
+    try {
+      setSelectedTodoIds(prevIds => [...prevIds, todoId]);
+
+      await deleteTodo(todoId).then(() => {
+        setTodosFromServer(
+          prevTodos => prevTodos.filter(todo => todo.id !== todoId),
+        );
+      });
+    } catch {
+      onError(ErrorType.UnableToDelete);
+    } finally {
+      setTimeout(
+        () => {
+          setSelectedTodoIds(ids => {
+            const idsCopy = [...ids];
+
+            idsCopy.splice(ids.indexOf(todoId), 1);
+
+            return idsCopy;
+          });
+
+          if (inputRef.current) {
+            inputRef.current.focus();
+          }
+        },
+        300,
+      );
+    }
+  };
+
+  const completedTodoIds = todosFromServer
+    .filter(todo => todo.completed)
+    .map(todo => todo.id);
+
+  const handleCompletedTodosDelition = async () => {
+    const processedIds: number[] = [];
+
+    try {
+      setSelectedTodoIds(prev => [...prev, ...completedTodoIds]);
+
+      completedTodoIds.forEach(async (todoId) => {
```

Comment:
`async` doesn't have any effect in `forEach`. If you want to proceed a batch of requests in the same time use `Promise.all`.
If you want to do them one by one you can do it with `forof`. Example with forof:
```javascript
for (const todoId of completedTodoIds) {
  await deleteTodo(todoId);
}
```

Reply:
> `async` doesn't have any effect in `forEach`. If you want to proceed a batch of requests in the same time use `Promise.all`. If you want to do them one by one you can do it with `forof`. Example with forof:
> 
> ```js
> for (const todoId of completedTodoIds) {
>   await deleteTodo(todoId);
> }
> ```

Not fixed, a few links about async in forEach:
https://gist.github.com/joeytwiddle/37d2085425c049629b80956d3c618971
https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop

Reply:
Not fixed 😢